### ISSUE TITLE
PHP symbol lookup error: undefined symbol

### ISSUE THREAD

**Issue by CharlotteDunois** [CopernicaMarketingSoftware/PHP-CPP#372]
Hello

I am trying to make my first PHP extension in C++ using PHP-CPP. So far I had a good experience. I was able to compile PHP-CPP and my extension, and PHP loads the extension without any errors (yay).

But when I'm trying to create a new instance of a class my extension declares, I get a symbol lookup error:
```
php: symbol lookup error: /usr/lib/php/20170718/erlpack.so: undefined symbol: _ZN15CharlotteDunois7Erlpack11__constructERN3Php10ParametersE
```
The source code of the extension is live on: https://github.com/CharlotteDunois/Erlpack-CPP

This is the PHP code I'm using:
```php
<?php
var_dump(extension_loaded('erlpack')); // prints bool(true)

$erlpack = new \CharlotteDunois\Erlpack\Erlpack(false);

var_dump($erlpack);
```

Output of php-config:
```
Usage: /usr/bin/php-config [OPTION]
Options:
  --prefix            [/usr]
  --includes          [-I/usr/include/php/20170718 -I/usr/include/php/20170718/main -I/usr/include/php/20170718/TSRM -I/usr/include/php/20170718/Zend -I/usr/include/php/20170718/ext -I/usr/include/php/20170718/ext/date/lib ]
  --ldflags           [-L/usr/lib/php/20170718  -L/usr/lib/x86_64-linux-gnu]
  --libs              [-lcrypt   -largon2 -lresolv -lcrypt -lz -lpcre -lrt -lm -ldl -lnsl  -lxml2 -lssl -lcrypto -lsodium -lcrypt -largon2 -largon2 -lcrypt ]
  --extension-dir     [/usr/lib/php/20170718]
  --include-dir       [/usr/include/php/20170718]
  --man-dir           [/usr/share/man]
  --php-binary        [/usr/bin/php7.2]
  --php-sapis         [cli fpm phpdbg ]
  --phpapi            [20170718]
  --configure-options [--includedir=/usr/include --mandir=/usr/share/man --infodir=/usr/share/info --disable-silent-rules --libdir=/usr/lib/x86_64-linux-gnu --libexecdir=/usr/lib/x86_64-linux-gnu --disable-maintainer-mode --disable-dependency-tracking --prefix=/usr --enable-cli --disable-cgi --disable-phpdbg --with-config-file-path=/etc/php/7.2/cli --with-config-file-scan-dir=/etc/php/7.2/cli/conf.d --build=x86_64-linux-gnu --host=x86_64-linux-gnu --config-cache --cache-file=/build/php7.2-9WZWiT/php7.2-7.2.3/config.cache --libdir=${prefix}/lib/php --libexecdir=${prefix}/lib/php --datadir=${prefix}/share/php/7.2 --program-suffix=7.2 --sysconfdir=/etc --localstatedir=/var --mandir=/usr/share/man --disable-all --disable-debug --disable-rpath --disable-static --with-pic --with-layout=GNU --without-pear --enable-filter --with-openssl=yes --with-password-argon2=/usr --with-pcre-regex=/usr --enable-hash --with-mhash=/usr --enable-libxml --enable-session --with-sodium --with-system-tzdata --with-zlib=/usr --with-zlib-dir=/usr --enable-dtrace --enable-pcntl --with-libedit=shared,/usr build_alias=x86_64-linux-gnu host_alias=x86_64-linux-gnu CFLAGS=-g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -O2 -Wall -pedantic -fsigned-char -fno-strict-aliasing -g]
  --version           [7.2.3-1+ubuntu16.04.1+deb.sury.org+1]
  --vernum            [70203]
```

g++ -v:
```
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.9' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.9) 
```

OS: Linux Mint 18

If you need some more information, please do not hesitate to ask.

I am not very experienced at C++, so if there is anything wrong, please tell me. It's a learning experience I want to make. Only if you learn something, you can improve yourself. :)

Best Regards

**Comment by sjinks** [comment]
The bug is in your Makefile:

```
SOURCES	= $(wildcard *.cpp)
```

This adds to `SOURCES` only `.cpp` files found in the same directory as your `Makefile`, and the files where your classes are actually implemented are never linked in.

You probably need something like this:

```Makefile
SOURCES = $(wildcard *.cpp) $(wildcard src/*.cpp)
```

**Comment by CharlotteDunois** [comment]
Thanks for this tip! This raised a few more compilation errors which I fixed, but the initial error is gone.

I've got a new error now when I try to use the method `decode`:
```
PHP Fatal error:  Cannot make a variable undefined
```
I've looked through the PHP-CPP code and it looks like that `Type::setValue` triggers this on `Type::UNDEFINED`.
I don't know when that function gets called and where - probably somewhere in phpcpp itself.

I am trying to figure it out, but maybe you know the solution to that error too?
